okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+                                        + "server-side QoS condition: 503", exception));
+                            } else {
+                                log.debug("Rescheduling call after backoff",
+                                        SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                                scheduleExecution(callback, backoff.get());
+                            }
+                            return null;
+                        }
+                    });
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
+                if (httpError.isPresent()) {
+                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    return;
+                }
+
+                // Catch-all: handle all other responses
+                Optional<IOException> ioException = catchAllExceptionHandler.handle(response);
+                if (ioException.isPresent()) {
+                    callback.onFailure(call, ioException.get());
+                    return;
+                }
+
+                callback.onFailure(call, new IOException("Failed to handle request, this is an http-remoting bug."));
+            }
+        });
+    }
+
+    @Override
+    public Response execute() throws IOException {
+        SettableFuture<Response> future = SettableFuture.create();
+        enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                future.setException(exception);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                future.set(response);
+            }
+        });
+
+        try {
+            return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Call was interrupted during execution");
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof IoRemoteException) {
+                throw ((IoRemoteException) e.getCause()).getWrappedException();
+            } else if (e.getCause() instanceof IOException) {
+                throw (IOException) e.getCause();
+            } else {
+                throw new IOException("Failed to execute call", e);
+            }
+        } catch (TimeoutException e) {
+            throw new IOException("Call timed out after: " + timeout.toString(), e);
+        }
+    }
+
+    @Override
+    public void enqueue(Callback callback) {
+        doEnqueueAndHandleErrorsAndRetries(callback);
+    }
+
+    @Override
+    public RemotingOkHttpCall doClone() {
```

Comment:
@j-baker objected to these doClone methods before... can we just make ForwardingCall an interface with lots of `default` methods, missing just `clone` and `getDelegate` implementations. Gets rid of a bit of inheritance :)

Reply:
future PR. https://github.com/palantir/http-remoting/issues/627